hs_core/hydroshare/utils.py

```diff
@@ -359,3 +364,47 @@ def resource_file_add_process(resource, files, user, extract_metadata=False, **k
     resource_modified(resource, user)
     return resource_file_objects
 
+
+def add_file_to_resource(resource, f):
+    """
+    Add a ResourceFile to a Resource.  Adds the 'format' metadata element to the resource.
+    :param resource: Resource to which file should be added
+    :param f: File-like object to add to a resource
+    :return: The identifier of the ResourceFile added.
+    """
+    ret = ResourceFile.objects.create(content_object=resource,
+                                      resource_file=File(f) if not isinstance(f, UploadedFile) else f)
+    # add format metadata element if necessary
+    file_format_type = get_file_mime_type(f.name)
+    if file_format_type not in [mime.value for mime in resource.metadata.formats.all()]:
+        resource.metadata.create_element('format', value=file_format_type)
+
+
+class ZipContents(object):
+    def __init__(self, zip_file):
+        self.zip_file = zip_file
+
+    def black_list_path(self, file_path):
+        return file_path.startswith('__MACOSX/')
+
+    def black_list_name(self, file_name):
+        return file_name == '.DS_Store'
+
+    def get_files(self):
+        logger = logging.getLogger('django')
+        temp_dir = tempfile.mkdtemp()
+        try:
+            for name_path in self.zip_file.namelist():
+                if not self.black_list_path(name_path):
+                    name = os.path.basename(name_path)
+                    if name != '':
+                        if not self.black_list_name(name):
+                            self.zip_file.extract(name_path, temp_dir)
+                            file_path = os.path.join(temp_dir, name_path)
```

Comment:
Should experiment with deleting file_path (if set) at the beginning of the loop as the file extracted in the previous iteration is no longer needed at that point.  This will reduce the amount of temporary disk space needed to the size of the largest file in the zip archive, instead of the size of all of the files in the archive.